remove some unused Conv constructors

CarloLucibello opened the pull request:
I think these constructors can go, I suspect no one is using them and they weren't even tested. 
Docstrings need some love, I won't do it here in order to avoid conflicts with #1391 

As a general rule, we consistently provide `Layer(W, b)` constructors, so no need to also have `Layer(; weight=W, bias=b)` for an arbitrary subset of the layers. 

### PR Checklist

- [ ] Tests are added
- [x] Entry in NEWS.md
- [ ] Documentation, if applicable
- [ ] Final review from `@dhairyagandhi96` (for API changes).

mcabbott commented:
To be clear, the constructors this deletes are the all-keyword ones. Keeping `Conv((3,3), 5=>7)` and `Conv(w, b; kw...)`.

CarloLucibello commented:
@DhairyaLGandhi this is waiting for approval

DhairyaLGandhi commented:
This doesn't remove any of the Dense constructors afaict so could update the news entry, and changes style which I would prefer to keep as the original.

Other than than, are we gaining much by removing the constructors?

CarloLucibello commented:
> Other than than, are we gaining much by removing the constructors?

no big deal, just some polishing and consistency improvement

DhairyaLGandhi commented on src/deprecations.jl:
```diff
@@ -3,3 +3,12 @@
+@deprecate Conv(; weight,  bias, σ=identity, kws...) Conv(weight, bias, σ; kws...) 
+@deprecate ConvTranspose(; weight, bias, σ=identity, kws...) ConvTranspose(weight, bias, σ; kws...) 
+@deprecate DepthwiseConv(; weight, bias, σ=identity, kws...) DepthwiseConv(weight, bias, σ; kws...) 
+
```
extra lines

DhairyaLGandhi commented on src/layers/basic.jl:
```diff
@@ -116,8 +116,8 @@ end
-function Dense(in::Integer, out::Integer, σ = identity;
-               initW = glorot_uniform, initb = zeros, bias=true)
+function Dense(in::Integer, out::Integer, σ=identity;
+               initW=glorot_uniform, initb=zeros, bias=true)
```
changes style - please use the consistent styling for all of Flux/ Zygote. I can appreciate having personal preferences, but I'd prefer to remain with how we have developed it thus far

CarloLucibello replied:
Notice that the spacing was inconsistent before. And that no blanks around `=` is not just my personal preference, it is widely adopted standard, see julia Base or https://github.com/invenia/BlueStyle#whitespace. I'd like to align our codebase to that

DhairyaLGandhi replied:
Even if it's not personal preference, it's unwieldy to change these things around if we have an established pattern we use.

DhairyaLGandhi replied:
Unless there's any real reason to do so, we can move ahead with the established pattern and revisit if we ever need to change it, and do it consistently then

mcabbott replied:
It lines it up with the pattern of `Conv`. And getting eltypes wrong seems to be about the most common silent performance bug.

CarloLucibello replied:
It's not really clear there is an established pattern. at present, there isn't much consistence, see the Losses module, the dropout function, normalize,....

DhairyaLGandhi replied:
We should fix it there as well, in that case.

CarloLucibello commented:
@DhairyaLGandhi are we good here?

CarloLucibello commented:
bump

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -356,25 +342,28 @@ end
 """
-function DepthwiseConv(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ = identity;
-                      stride = 1, pad = 0, dilation = 1) where {T,N}
+function DepthwiseConv(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ=identity;
+                      stride=1, pad=0, dilation=1) where {T,N}
```
```suggestion
function DepthwiseConv(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ = identity;
                      stride = 1, pad = 0, dilation = 1) where {T,N}
```

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -245,31 +238,24 @@ end
-function ConvTranspose(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ = identity;
-                      init = glorot_uniform, stride = 1, pad = 0, dilation = 1,
-                      weight = convfilter(k, reverse(ch), init = init), bias = true) where N
+function ConvTranspose(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ=identity;
+                      init=glorot_uniform, stride=1, pad=0, dilation=1,
+                      weight=convfilter(k, reverse(ch), init =init), bias=true) where N
```
```suggestion
function ConvTranspose(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ = identity;
                      init = glorot_uniform, stride = 1, pad = 0, dilation = 1,
                      weight=convfilter(k, reverse(ch), init = init), bias = true) where N
```

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -245,31 +238,24 @@ end
 """
-function ConvTranspose(w::AbstractArray{T,N}, b::Union{Bool,AbstractVector{T}}, σ = identity;
-                      stride = 1, pad = 0, dilation = 1) where {T,N}
+function ConvTranspose(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ=identity;
+                      stride=1, pad=0, dilation=1) where {T,N}
```
```suggestion
function ConvTranspose(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ = identity;
                      stride = 1, pad = 0, dilation = 1) where {T,N}
```

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -129,18 +128,20 @@ julia> params(c1stride) |> length  # no trainable bias
-function Conv(;weight::AbstractArray{T,N}, bias::Union{Bool, Zeros, AbstractVector{T}},
-              activation = identity, stride = 1, pad = 0, dilation = 1) where {T,N}
-  Conv(weight, bias, activation, stride = stride, pad = pad, dilation = dilation)
+function Conv(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ=identity;
+            init=glorot_uniform,  stride=1, pad=0, dilation=1,
+            weight = convfilter(k, ch; init), bias=true) where N
+
```
```suggestion
function Conv(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ = identity;
            init = glorot_uniform,  stride = 1, pad = 0, dilation = 1,
            weight = convfilter(k, ch; init), bias = true) where N

```

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -129,18 +128,20 @@ julia> params(c1stride) |> length  # no trainable bias
 """
-function Conv(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ = identity;
-              stride = 1, pad = 0, dilation = 1) where {T,N}
+function Conv(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ=identity;
+              stride=1, pad=0, dilation=1) where {T,N}
```
```suggestion
function Conv(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ = identity;
              stride = 1, pad = 0, dilation = 1) where {T,N}
```

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -356,25 +342,28 @@ end
-function DepthwiseConv(;weight::AbstractArray{T,N}, bias::Union{Bool, Zeros, AbstractVector{T}},
-                      activation = identity, stride = 1, pad = 0, dilation = 1) where {T,N}
-  DepthwiseConv(weight, bias, activation, stride = stride, pad = pad, dilation = dilation)
+function DepthwiseConv(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ=identity;
+                init=glorot_uniform, stride=1, pad=0, dilation=1,
+                weight=depthwiseconvfilter(k, ch, init=init), bias=true) where N
+  @assert ch[2] % ch[1] == 0 "Output channels must be integer multiple of input channels"
+  return DepthwiseConv(weight, bias, σ; stride, pad, dilation)
```
```suggestion
function DepthwiseConv(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ = identity;
                init = glorot_uniform, stride = 1, pad = 0, dilation = 1,
                weight = depthwiseconvfilter(k, ch, init = init), bias = true) where N
  @assert ch[2] % ch[1] == 0 "Output channels must be integer multiple of input channels"
  return DepthwiseConv(weight, bias, σ; stride, pad, dilation)
```

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -466,31 +438,24 @@ end
 """
-function CrossCor(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ = identity;
-                  stride = 1, pad = 0, dilation = 1) where {T,N}
+function CrossCor(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ=identity;
+                  stride=1, pad=0, dilation=1) where {T,N}
```
```suggestion
function CrossCor(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, σ = identity;
                  stride = 1, pad = 0, dilation = 1) where {T,N}
```

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -466,31 +438,24 @@ end
-function CrossCor(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ = identity;
-                  init = glorot_uniform, stride = 1, pad = 0, dilation = 1,
-                  weight = convfilter(k, ch, init = init), bias = true) where N
+function CrossCor(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ=identity;
+                  init=glorot_uniform, stride=1, pad=0, dilation=1,
+                  weight=convfilter(k, ch, init=init), bias = true) where N
```
```suggestion
function CrossCor(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ = identity;
                  init = glorot_uniform, stride = 1, pad = 0, dilation = 1,
                  weight = convfilter(k, ch, init = init), bias = true) where N
```

DhairyaLGandhi commented on src/layers/basic.jl:
```diff
@@ -116,8 +116,8 @@ end
 
-function Dense(in::Integer, out::Integer, σ = identity;
-               initW = glorot_uniform, initb = zeros, bias=true)
+function Dense(in::Integer, out::Integer, σ=identity;
+               initW=glorot_uniform, initb=zeros, bias=true)
```
```suggestion
function Dense(in::Integer, out::Integer, σ = identity;
               initW = glorot_uniform, initb = zeros, bias = true)
```

DhairyaLGandhi commented on NEWS.md:
```diff
@@ -5,11 +5,12 @@
 * Excise datasets in favour of other providers in the julia ecosystem.
 * other new features and bug fixes (see GitHub's releases page)
 * Added option to set `bias` to [false](https://github.com/FluxML/Flux.jl/pull/1379) to eliminating `bias` from being trained.
+* Some not commonly used constructors are removed from [`Conv layers`](https://github.com/FluxML/Flux.jl/pull/1379)).
```
```suggestion
* Removed kwarg only constructors for [`convolutional layers`](https://github.com/FluxML/Flux.jl/pull/1379)).
```

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -259,17 +251,11 @@ function ConvTranspose(w::AbstractArray{T,N}, b::Union{Bool,AbstractVector{T}},
 function ConvTranspose(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ = identity;
                       init = glorot_uniform, stride = 1, pad = 0, dilation = 1,
-                      weight = convfilter(k, reverse(ch), init = init), bias = true) where N
+                      weight=convfilter(k, reverse(ch), init = init), bias = true) where N
```
```suggestion
                      weight = convfilter(k, reverse(ch), init = init), bias = true) where N
```

DhairyaLGandhi commented on src/layers/conv.jl:
```diff
@@ -138,9 +137,11 @@ function Conv(w::AbstractArray{T,N}, b::Union{Bool, Zeros, AbstractVector{T}}, 
-              activation = identity, stride = 1, pad = 0, dilation = 1) where {T,N}
-  Conv(weight, bias, activation, stride = stride, pad = pad, dilation = dilation)
+function Conv(k::NTuple{N,Integer}, ch::Pair{<:Integer,<:Integer}, σ = identity;
+            init = glorot_uniform,  stride = 1, pad = 0, dilation = 1,
```
```suggestion
            init = glorot_uniform, stride = 1, pad = 0, dilation = 1,
```

DhairyaLGandhi commented:
Seems like we either need to fix the docs or these constructors were actually used

CarloLucibello commented:
I fixed some rebase mess I created, squashed the commits, fixed documentation and deprecation issues.
This should be finally ready to merge when CI goes green

CarloLucibello commented:
@DhairyaLGandhi now should be ready ready :)

CarloLucibello commented:
@DhairyaLGandhi bump

DhairyaLGandhi commented:
bors r+

CarloLucibello commented:
@DhairyaLGandhi needs a review approval, otherwise bors won't be able to merge given current github's settings

CarloLucibello commented:
rebased

DhairyaLGandhi commented:
bors r+

CarloLucibello commented:
Ahahh ah, thank you bors
